asf_heat_pump_suitability/getters/get_dno_datasets.py

```diff
@@ -68,35 +69,44 @@ def generate_enw_gdf() -> gpd.GeoDataFrame:
     enw["firm_capacity_mva"] = (
         enw["maximum_demand_mva_per_primary_substation"] + enw["headroom_mva"]
     )
-    enw["operator"] = "ENW"
     enw["geopoint"] = enw["geopoint"].apply(parse_binary_geometry)
 
-    # Create GeoDataFrame and perform spatial join with shape data
+    # Distribution areas are given by primary group rather than individual primary substations
+    # Substation data doesn't contain information on primary group, so we need to perform a spatial join
     enw_gdf = gpd.GeoDataFrame(enw, geometry="geopoint", crs=CRS)
     # polygon geometry of substation distribution area
     enw_shape["geo_shape"] = enw_shape["geo_shape"].apply(parse_binary_geometry)
-    # point geometry of substation location
-    enw_shape["geo_point_2d"] = enw_shape["geo_point_2d"].apply(parse_binary_geometry)
     enw_shape_gdf = gpd.GeoDataFrame(enw_shape, geometry="geo_shape", crs=CRS)
-    enw_df = gpd.sjoin(enw_gdf, enw_shape_gdf, how="right", predicate="intersects")
+    enw_df = gpd.sjoin(enw_gdf, enw_shape_gdf, how="right", predicate="within")
     enw_df = enw_df.dropna(subset=["index_left"])
 
-    # Prepare final dataframe
-    enw_df["substation_id"] = enw_df["pry_group"] + "_" + enw_df["substation"]
-    return enw_df[
-        [
-            "substation_id",
-            "firm_capacity_mva",
-            "maximum_demand_mva_per_primary_substation",
-            "geo_shape",
-            "operator",
+    # Aggregate capacity and demand data to primary group level
+    enw_df = (
+        enw_df.groupby(["pry_group", "geo_shape"])[
+            ["firm_capacity_mva", "maximum_demand_mva_per_primary_substation"]
         ]
-    ].rename(
-        columns={
-            "substation_id": "id",
-            "maximum_demand_mva_per_primary_substation": "peak_demand_mva",
-            "geo_shape": "geo_shape",
-        }
+        .sum()
+        .reset_index()
+    )
+    enw_df["operator"] = "ENW"
+    return gpd.GeoDataFrame(
+        enw_df[
+            [
+                "pry_group",
+                "firm_capacity_mva",
+                "maximum_demand_mva_per_primary_substation",
+                "geo_shape",
+                "operator",
+            ]
+        ].rename(
+            columns={
+                "substation_id": "id",
+                "maximum_demand_mva_per_primary_substation": "peak_demand_mva",
+                "geo_shape": "geo_shape",
```

Comment:
Should `"substation_id"` in this dict be `"pry_group"` now? Also, i dont think you need to rename `geo_shape`

Reply:
Later, when the DNO gdfs are concatenated in `grid_capacity.generate_substations_gdf` we end up with `pry_group` col and no `id` col. The `pry_group` col has nan values for the other DNOs.